homeassistant/components/obihai/sensor.py

```diff
@@ -0,0 +1,99 @@
+"""Support for Obihai Sensors."""
+import logging
+
+from datetime import timedelta
+import voluptuous as vol
+
+from pyobihai import PyObihai
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=5)
+
+DOMAIN = "Obihai"
+DEFAULT_USERNAME = "admin"
+DEFAULT_PASSWORD = "admin"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_HOST): cv.string,
+        vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+        vol.Optional(CONF_PASSWORD, default=DEFAULT_PASSWORD): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Obihai sensor platform."""
+
+    username = config.get(CONF_USERNAME, None)
+    password = config.get(CONF_PASSWORD, None)
+    host = config.get(CONF_HOST, None)
+
+    sensors = []
+
+    pyobihai = PyObihai()
+
+    services = pyobihai.get_state(host, username, password)
+
+    line_services = pyobihai.get_line_state(host, username, password)
+
+    for key in services:
+        sensors.append(ObihaiServiceSensors(host, username, password, key))
+
+    for key in line_services:
+        sensors.append(ObihaiServiceSensors(host, username, password, key))
+
+    add_devices(sensors)
+
+
+class ObihaiServiceSensors(Entity):
+    """Get the status of each Obihai Lines."""
+
+    def __init__(self, host, username, password, service_name):
+        """Initialize monitor sensor."""
+        self._host = host
+        self._username = username
+        self._password = password
+        self._service_name = service_name
+        self._state = None
+        self._name = "{} {}".format(DOMAIN, self._service_name)
+        self._pyobihai = PyObihai()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    def update(self):
+        """Update the sensor."""
+        services = self._pyobihai.get_state(self._host, self._username, self._password)
+
+        if self._service_name in services:
+            if services[self._service_name] is None:
+                self._state = STATE_UNKNOWN
+            else:
+                self._state = services[self._service_name]
```

Comment:
What type of state values does the different entities return?

Reply:
There are a few sensors the users will get:

- Obihai service status (Normal, Down or other states from Obihais network)
- Sensors for each phone port in use (On Hook, Off Hook and Ringing)
- Sensors for last caller name and number (this is also the current incoming call)
- Sensor if the device requires a reboot
- Sensor for each configured service (0 for no calls, 1 for a call and 2 for call waiting/3way calling)
- An uptime sensor that does not appear to be formatted properly for date time as it arbitrarily lists the days (ex: `2 Days 18:59:10 (3)`)

Reply:
Will the update sensor state change just because time passes? If so we should convert it to an absolute utc timestamp.

Reply:
The uptime sensor does update as time passes, but it resets whenever the device is rebooted.  I honestly expected mine to be higher than 2 days but I guess it rebooted on its own :)

Reply:
Ok. We should try to convert it to an absolute utc timestamp and set `device_class` property to `"timestamp"` for that entity.

Reply:
Ok I am not really sure where to begin to convert the output I listed to UTC :(  I think the format I am retrieving is incorrect.

```
current=" 0 Days  0:30:50    (3)"
```

Reply:
If we pass the days, hours, minutes and seconds to `datetime.timedelta` would it be correct?

https://docs.python.org/3/library/datetime.html#datetime.timedelta

Reply:
Then we can just subtract that timedelta from the current datetime to get another datetime, which we use as timestamp.

Reply:
@MartinHjelmare ok if I am to understand this correctly I just need to return the state in a `timedelta` format?

```
state = timedelta(days=days, hours=h, minutes=m, seconds=s)
```
this is the output I get now which is slightly different from earlier:

```
2 days, 20:30:37
```
Then I can set the device class?

Reply:
No, we need to convert the time delta to an utc iso formatted datetime. Take `dt_util.utcnow` and subtract the timedelta. This should get us a datetime in the past when the device started. Then return the iso formatted string of that time as state.

Reply:
It's weird that the output changes 2 days within the hour.

Reply:
@MartinHjelmare sorry you can consider that a copy and paste error.  I had previously saved the output I was getting and pasted that here.  The time I posted here: https://github.com/home-assistant/home-assistant/pull/26537#discussion_r322439961

and here: https://github.com/home-assistant/home-assistant/pull/26537#discussion_r322473473

are actual live values :) sorry about that

Reply:
@MartinHjelmare with this change the sensor will then just display the datetime of last reboot correct? So I should change the name from uptime to something else like last reboot?

Reply:
By setting the device_class to timestamp, lovelace will display the state as relative time in the entities card.

Still, to name the state properly it's probably a good idea to change name.

Reply:
Ok made some progress this is new state output `2019-09-07T01:26:19.335922` I will go ahead and apply a device class for this sensor and rename it accordingly

Reply:
disregard last commit, made a mistake :)

Reply:
ok should be good to go now :)